lib/compiler.js

```diff
@@ -288,63 +289,63 @@ function getFileDependencies (compiler) {
 }
 
 /**
- * @type {WeakMap<WebpackCompilation, WeakMap<HtmlWebpackChildCompiler, boolean>>}}
+ * @type {WeakMap<object, Snapshot>}
  */
-const hasOutdatedCompilationDependenciesMap = new WeakMap();
+const snapshotMap = new WeakMap();
+const fileDepependenciesKeyRef = {};
 /**
- * Returns `true` if the file dependencies of the current childCompiler
- * for the given mainCompilation are outdated.
+ * Calls the callback with `true` if the file dependencies of the current childCompiler
+ * for the given mainCompilation are valid.
  *
- * Uses the `hasOutdatedCompilationDependenciesMap` cache if possible.
+ * Uses the `snapshotMap` cache if possible.
  *
  * @param {WebpackCompilation} mainCompilation
- * @returns {boolean}
+ * @param callback
  */
-function hasOutDatedTemplateCache (mainCompilation) {
+function hasOutDatedTemplateCache (mainCompilation, callback) {
   const childCompiler = getChildCompiler(mainCompilation.compiler);
-  /**
-   * @type {WeakMap<HtmlWebpackChildCompiler, boolean>|undefined}
-   */
-  let hasOutdatedChildCompilerDependenciesMap = hasOutdatedCompilationDependenciesMap.get(mainCompilation);
-  // Create map for childCompiler if none exist
-  if (!hasOutdatedChildCompilerDependenciesMap) {
-    hasOutdatedChildCompilerDependenciesMap = new WeakMap();
-    hasOutdatedCompilationDependenciesMap.set(mainCompilation, hasOutdatedChildCompilerDependenciesMap);
+
+  const keyStr = JSON.stringify([...childCompiler.fileDependencies].sort());
+  // get reference for key-strings to use it with WeakMap:
+  let key = fileDepependenciesKeyRef[keyStr];
+  if (!key) {
+    key = {};
+    fileDepependenciesKeyRef[keyStr] = key;
   }
-  // Try to get the `checkChildCompilerCache` result from cache
-  let isOutdated = hasOutdatedChildCompilerDependenciesMap.get(childCompiler);
-  if (isOutdated !== undefined) {
-    return isOutdated;
+
+  function updateSnapshot () {
+    mainCompilation.fileSystemInfo.createSnapshot(
+      Date.now(),
+      childCompiler.fileDependencies,
+      childCompiler.contextDependencies,
+      childCompiler.missingDependencies,
+      null,
+      (err, snapshot) => {
+        if (err) {
+          return callback(err);
+        }
+        snapshotMap.set(key, snapshot);
+      }
+    );
   }
-  // If `checkChildCompilerCache` has never been called for the given
-  // `mainCompilation` and `childCompiler` combination call it:
-  isOutdated = isChildCompilerCacheOutdated(mainCompilation, childCompiler);
-  hasOutdatedChildCompilerDependenciesMap.set(childCompiler, isOutdated);
-  return isOutdated;
-}
 
-/**
- * Returns `true` if the file dependencies of the given childCompiler are outdated.
- *
- * @param {WebpackCompilation} mainCompilation
- * @param {HtmlWebpackChildCompiler} childCompiler
- * @returns {boolean}
- */
-function isChildCompilerCacheOutdated (mainCompilation, childCompiler) {
```

Comment:
why did you remove this part?

Reply:
Comes from PutziSan's changes, I'll need to take a look.

Reply:
It is because the checking went from being sync to async

Reply:
Oh, I saw wrong. isChildCompilerCacheOutdated needs to go back in.